### ISSUE TITLE
Support stock iOS dynamic font scaling

### ISSUE THREAD

**Issue by samskiter** [material-components/material-components-ios#4597]
## Feature Request/Suggestion

We are developing all of our UI components for content to support stock iOS dynamic font scaling. We would therefore like the MDC components that we use alongside them to scale with the same scaling curves. So rather than using `mdc_setAdjustsFontForContentSizeCategory` we'd like to use `setAdjustsFontForContentSizeCategory` directly.

On a related ticket (https://github.com/material-components/material-components-ios/issues/4346) we've spoken a little about iOS versions being tricky in implementing this, but I just wanted to mention my take on the use-cases here:

1. If we were supporting iOS < 10 and wanted to support dynamic type: we would probably have opted to use the MDC Typography component in our own UI components too as they give an easy, free scaling curve.
2. As we are supporting iOS >= 10 we just want to use the stock scaling curves and have no need for the MDC curves.

So I think if you were to add a method like `mdc_setSystemAdjustsFontForContentSizeCategory` to allow using the system scaling, you could make that only available from iOS 10+ and not worry about a fallback.

## Additional context

Some discussion here
https://github.com/material-components/material-components-ios/issues/4346

We will add some information about a specific Bug related to this in the MDCButton component + stock dynamic type.

---
#### Internal data
- Associated internal bug: [b/117177514](http://b/117177514)

**Comment by babakbo** [comment]
The current implementation is problematic if system dynamic font is being used.

Consider the following code:

```swift
let button = MDCButton()
button.setTitle("Button", for: .normal)
button.setTitleFont(UIFont.preferredFont(forTextStyle: .body), for: .normal)
button.titleLabel?.adjustsFontForContentSizeCategory = true
```

This will present an MDCButton which will automatically resize based on (and consistent with) the built-in font resizing (which is good).
However, as soon as user taps on the button, the button resizes to its original size (bad!).
See the attached video for a better understanding of the issue.

The root cause is the MDCButton.updateTitleFont. At each state change this method tries to assign the new font to the titleLabel. In this example, we have only one font, however, the font we keep in _fonts is not scaled (even fonts that are defined with preferredFontForTextStyle do not scale automatically without the view's help.) . Therefore reassigning it to the titleLabel resets the label size to the original size (which is incorrect with the given dynamicType).

To fix this, we should replace 

`self.titleLabel.font = font` 

with

```objc
if (@available(iOS 11.0, *)) {
  self.titleLabel.font = [fontMetrics scaledFontForFont:font];
} else {
  self.titleLabel.font = font;
}
```
where fontMetric ideally should be a property of type UIFontMetrics that users can set.

[MDCButtonDynamicType.zip](https://github.com/material-components/material-components-ios/files/2211543/MDCButtonDynamicType.zip)

**Comment by ianegordon** [comment]
Unfortunately, Apple hasn't exposed the hooks we need to directly interact with their Dynamic Type system.  Although we can now attach a custom font to one of Apple's font styles, we **cannot** attach our own curves.  (As of iOS 12)

Our upcoming implementation of Material Dynamic Type 2.0 will make the API closer to what you are asking for, but unfortunately, there is no solution :(

Feel free to reach out directly if you'd like to discuss the details.
